src/cpu/o3/bac.cc

```diff
@@ -0,0 +1,1069 @@
+    }
+    if ((bacStatus[tid] == FTQLocked) && !ftq->isLocked(tid)) {
+
+        DPRINTF(BAC, "[tid:%i] FTQ not locked anymore -> Running\n", tid);
+        bacStatus[tid] = Running;
+        return true;
+    }
+
+    // Check if the FTQ became free in that cycle.
+    if ((bacStatus[tid] == FTQFull) && !ftq->isFull(tid)) {
+
+        DPRINTF(BAC, "[tid:%i] FTQ not full anymore -> Running\n", tid);
+        bacStatus[tid] = Running;
+        return true;
+    }
+
+    if (bacStatus[tid] == Squashing) {
+
+        // Switch status to running after squashing FTQ and setting the PC.
+        DPRINTF(BAC, "[tid:%i] Done squashing, switching to running.\n", tid);
+        bacStatus[tid] = Running;
+        return true;
+    }
+
+    // Now all stall/squash conditions are checked.
+    // Attempt to run the BAC if not already running.
+    if (ftq->isValid(tid) &&
+            ((bacStatus[tid] == Idle) || (bacStatus[tid] == Blocked))) {
+
+        DPRINTF(BAC, "[tid:%i] Attempt to run\n", tid);
+        bacStatus[tid] = Running;
+        return true;
+    }
+
+    // If we've reached this point, we have not gotten any signals that
+    // cause BAC to change its status.  BAC remains the same as before.
+    return false;
+}
+
+
+
+
+void
+BAC::squashBpuHistories(ThreadID tid)
+{
+    if (!decoupledFrontEnd) return;
+
+    DPRINTF(BAC, "%s(tid:%i): FTQ sz: %i\n", tid, __func__, ftq->size(tid));
+
+    unsigned n_fts = ftq->size(tid);
+    if (n_fts == 0) return;
+
+    // Iterate over the FTQ in reverse order to
+    // revert all predictions made.
+    ftq->forAllBackward(tid,
+        [this, tid](FetchTargetPtr &ft)
+        {
+            if (ft->bpu_history) {
+                auto hist = static_cast<BPredUnit::PredictorHistory*>
+                                                    (ft->bpu_history);
+                bpu->squashHistory(tid, hist);
+                assert(hist == nullptr);
+                ft->bpu_history = nullptr;
+            }
+        });
+}
+
+void
+BAC::squash(const PCStateBase &new_pc, ThreadID tid)
+{
+    if (!decoupledFrontEnd) return;
+
+    DPRINTF(BAC, "[tid:%i] Squashing FTQ.\n", tid);
+
+    // Set status to squashing.
+    bacStatus[tid] = Squashing;
+
+    // Set the new PC
+    set(bacPC[tid], new_pc);
+
+    // Then squash all fetch targets
+    ftq->squash(tid);
+}
+
+
+void
+BAC::tick()
+{
+    bool activity = false;
+    bool status_change = false;
+
+    std::list<ThreadID>::iterator threads = activeThreads->begin();
+    std::list<ThreadID>::iterator end = activeThreads->end();
+
+if (decoupledFrontEnd) {
+    // FDP ---------------------------------------------
+    // In the decoupled frontend the BAC stage is active as all
+    // others. Its main purpose is generate fetch targets by using
+    // the branch predction unit.
+
+    while (threads != end) {
+        ThreadID tid = *threads++;
+
+        // Check stall and squash signals first.
+        status_change |= checkSignalsAndUpdate(tid);
```

Comment:
Is this logical or bitwise or... I think it's supposed to be logical. I would make that explicit.

```suggestion
        status_change = status_change || checkSignalsAndUpdate(tid);
```